plugin/evm/acp176/acp176.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	bigExcess := new(big.Int).SetUint64(uint64(s.Gas.Excess))
+	bigPriceUpdateConversion := new(big.Int).SetUint64(uint64(priceUpdateConversion))
+	return fakeExponential(minBaseFee, bigExcess, bigPriceUpdateConversion)
+}
+
+// AdvanceTime increases the gas capacity and decreases the gas excess based on
+// the elapsed seconds.
+func (s *State) AdvanceTime(seconds uint64) {
+	targetPerSecond := s.Target()
+	maxPerSecond, err := safemath.Mul(TargetToMax, targetPerSecond)
+	if err != nil {
+		maxPerSecond = math.MaxUint64
+	}
+	maxCapacity, err := safemath.Mul(TimeToFillCapacity, maxPerSecond)
+	if err != nil {
+		maxCapacity = math.MaxUint64
+	}
+	s.Gas = s.Gas.AdvanceTime(
+		maxCapacity,
+		maxPerSecond,
+		targetPerSecond,
+		seconds,
+	)
+}
+
+// ConsumeGas decreases the gas capacity and increases the gas excess by
+// gasUsed + extraGasUsed. If the gas capacity is insufficient, an error is
+// returned.
+func (s *State) ConsumeGas(
+	gasUsed uint64,
+	extraGasUsed *big.Int,
+) error {
+	var err error
+	s.Gas, err = s.Gas.ConsumeGas(gas.Gas(gasUsed))
+	if err != nil {
+		return err
+	}
+	if extraGasUsed != nil {
+		if !extraGasUsed.IsUint64() {
+			return fmt.Errorf("%w: extraGasUsed (%d) exceeds MaxUint64",
+				gas.ErrInsufficientCapacity,
+				extraGasUsed,
+			)
+		}
+		s.Gas, err = s.Gas.ConsumeGas(gas.Gas(extraGasUsed.Uint64()))
+	}
+	return err
+}
+
+// UpdateTargetExcess updates the targetExcess to be as close as possible to the
+// desiredTargetExcess without exceeding the maximum targetExcess change.
+func (s *State) UpdateTargetExcess(desiredTargetExcess gas.Gas) {
+	previousTargetPerSecond := s.Target()
+	s.TargetExcess = targetExcess(s.TargetExcess, desiredTargetExcess)
+	newTargetPerSecond := s.Target()
+	s.Gas.Excess = scaleExcess(
+		s.Gas.Excess,
+		newTargetPerSecond,
+		previousTargetPerSecond,
+	)
+}
+
+// DesiredTargetExcess calculates the optimal desiredTargetExcess given the
+// desired target.
+//
+// This could be solved directly by calculating D * ln(desiredTarget / P) using
+// floating point math. However, it introduces inaccuracies. So, we use a binary
+// search to find the closest integer solution.
+func DesiredTargetExcess(desiredTarget gas.Gas) gas.Gas {
+	return gas.Gas(sort.Search(maxTargetExcess, func(targetExcessGuess int) bool {
+		state := State{
+			TargetExcess: gas.Gas(targetExcessGuess),
+		}
+		return state.Target() >= desiredTarget
+	}))
+}
+
+// fakeExponential approximates factor * e ** (numerator / denominator) using
+// Taylor expansion:
+// e^x = 1 + x/1! + x^2/2! + x^3/3! + ...
+func fakeExponential(factor, numerator, denominator *big.Int) *big.Int {
+	var (
+		output = new(big.Int)
+		accum  = new(big.Int).Mul(factor, denominator)
+	)
+	for i := 1; accum.Sign() > 0; i++ {
+		output.Add(output, accum)
+
+		accum.Mul(accum, numerator)
+		accum.Div(accum, denominator)
+		accum.Div(accum, big.NewInt(int64(i)))
+	}
+	return output.Div(output, denominator)
+}
```

Comment:
@ARR4N Specifically interested in your opinion on copying this from the eip-4844 package vs exporting the function in the eip-4844 package.

Reply:
I suppose one alternative could be defining an exported version and just calling the unexported version.

Reply:
I think it's safer to copy (with attribution). 4844 is only the beginning of a lot of work they're doing with respect to sharding so who knows how things might change 🤷